cpp/include/raft/neighbors/detail/ivf_pq_build.cuh

```diff
@@ -404,8 +404,11 @@ void train_per_subset(raft::resources const& handle,
   auto device_memory = resource::get_workspace_resource(handle);
 
   rmm::device_uvector<float> pq_centers_tmp(index.pq_centers().size(), stream, device_memory);
-  rmm::device_uvector<float> sub_trainset(n_rows * size_t(index.pq_len()), stream, device_memory);
-  rmm::device_uvector<uint32_t> sub_labels(n_rows, stream, device_memory);
+  // Subsampling the train set for codebook generation. Using similar subsampling strategy as train_per_cluster
+  size_t big_enough     = 256ul * std::max<size_t>(index.pq_book_size(), index.pq_dim());
```

Comment:
Per-cluster codebooks approach is quite different than the per-subspace, because there we train many more codebooks (`n_lists` - thousands of them). Because of that, we drastically reduce the training set size for each cluster. The hope is that each codebook is only used for a few thousand vectors at most (typical cluster size), the loss in recall should be not significant (the recall is lower, but I don't know whether this is due to aggressive subsampling).

Having said that, I'm really curious to see the figures - how big is the recall loss (if any). Two concerns:
  1. You always take the beginning of the larger set; in per-cluster mode this is not a big deal, because the clusters are many and their sizes are small and the pq_len-subvectors are sort-of "interleaved" when we reshape the data. In per-subspace mode the same strategy means we look only at the beginning of the whole training set, which sounds not optimal.
  2. Per-subspace mode is what we directly compare against original faiss implementation in the benchmarks. Could you please check if they use subsampling at this level too? We'd need to think how to continue to fairly compare the implementations.

Reply:
Thanks for the information. For (1), it sounds like we should do a premutation then select the first `pq_n_rows`, or do random sampling. 
For (2), I briefly looked up FAISS code, it seems they do subsampling when the input number of rows is more than `max_points_per_centroid` per cluster, where the default `max_points_per_centroid` is 256. Similar to what RAFT uses for `per_cluster` method. [link to FAISS code](https://github.com/facebookresearch/faiss/blob/main/faiss/Clustering.cpp#L302)
Given this, I feel we should implement (1), and benchmark more datasets. Would you agree?

Reply:
I've seen FAISS does the subsampling for the IVF-lists clustering in the code you've shown; we control this in [the benchmark wrapper code](https://github.com/rapidsai/raft/blob/80a48ca33c5d57d9b0597aacfe72fd0f87d40be6/cpp/bench/ann/src/faiss/faiss_gpu_wrapper.h#L199C7-L213). I'm not sure, however, if the same code is used to sub-sample further for the codebooks (or if there is any further subsampling at all). Is this the case?
What you're proposing here is to add two-level subsampling; for the [IVF lists](https://github.com/rapidsai/raft/blob/80a48ca33c5d57d9b0597aacfe72fd0f87d40be6/cpp/include/raft/neighbors/detail/ivf_pq_build.cuh#L1753) first, and then for the codebooks on top of that. I think we don't have this fine-grained control over FAISS indexes, so it's going to be harder to compare the implementations fairly. Perhaps, a way around this would be to add a new explicit parameter in raft IVF-PQ implementation, so that we can manually adjust the subsampling ratio for benchmarking purposes? It could also help with the potential degrading recall issue.

Let's see how much subsampling we actually do on a typical example.
Let's say `n_rows = 100M`, `pq_dim = 64`, `pq_bits = 8`, `n_lists = 50K`, `params.kmeans_trainset_fraction = 0.1`.
Then the trainset size after the initial subsampling is 10M, or ~200 points per list.

In the `per_cluster` mode, the list data is reshaped, so it becomes `200*pq_dim = 12800` training samples per codebook. This is still less than `256 * max(pq_dim, 2^pq_bits) = 65536`, so no further subsampling is done in this case.
In the `per_subspace` mode, we take a slice of data per codebook on the whole training set, hence 10M points. Reducing this to 65536 per codebook means an extra ~1/150 subsampling on top of the first-level subsampling, which sounds like a lot a lot. Hence, I'm concerned about the implications on the recall. Btw, the `pq_dim` is not relevant in the `big_enough` formula in the `per_subspace` mode, because we don't reshape the data (and hence don't multiply the sample size); I'd just keep it at `some_constant * index.pq_book_size()`.

Overall, I'd suggest the following:
  1. Add the codebook subsampling parameter to the IVF-PQ build parameters with a sensible default.
  2. Maybe change the way you sample for codebooks from taking the first `pq_n_rows` to taking the same number of rows in an interleaved fashion?
  3. If (2) is not enough, maybe it's better to finally introduce permutation sampling to [the first-level sampling](https://github.com/rapidsai/raft/blob/80a48ca33c5d57d9b0597aacfe72fd0f87d40be6/cpp/include/raft/neighbors/detail/ivf_pq_build.cuh#L1764) rather than doing permutation for every subspace codebook.
  4. (for bonus points) Check if FAISS does subsampling on both levels or just once and see how we can compare fairly.

Reply:
Thanks @achirkin  for the suggestions! What matters is how good the codebooks are in the end. We only train max 256 cluster centers for each pq_dim, and the goal here is to find what is the right number of points to do that. (It is probably way less than the 10M points used here). I like the idea for adding a knob for control to `ivf_pq::index_params`.

Reply:
Regarding how to select the vectors, I agree with Artem that we need to add subsampling. Currently we have two variants:
- Select every `r`-th element, where `r` is the subsample ratio. This is currently used in `ivf_pq::build` https://github.com/rapidsai/raft/blob/80a48ca33c5d57d9b0597aacfe72fd0f87d40be6/cpp/include/raft/neighbors/detail/ivf_pq_build.cuh#L1766
- random subsampling, we have a code example here https://github.com/rapidsai/raft/blob/80a48ca33c5d57d9b0597aacfe72fd0f87d40be6/cpp/template/src/common.cuh#L73-L94

I think the random sampling shall be used as subsampling method for IVF cluster training. Ideally that should be a separate PR.

After that, it shall be sufficient to select every n-th element for codebook training (similar to what we do now with `cudaMemcpy2DAsync` for the cluster training).

Reply:
Thanks, Tamas, for providing the examples!
Note, in this code (train_per_subset) select-every-rth-element subsampling of the data can be implemented just by modifying the stride parameter of the linalg::gemm (the labels will need another cudaMemcpy2DAsync though). However, if we do the random sampling with permutation initially (for the IVF cluster training), we can safely leave this code as-is, right?

Reply:
That is right.

Reply:
Thanks @tfeher @achirkin for the comments. I think this PR depends on the random subsampling PR that Tamas is proposing. After that is merged, we need to 

1. Modify the `big_enough` formula by removing `pq_dim` dependency 
2. Add knob to control the amount of subsampling for PQ code generation 
3. Since random sampling with premutation at the first level is going to be supported, we could just select the first certain number of rows of dataset for PQ codebook training. Selecting every n-th element is not necessary 
4. Benchmark with real dataset, look into whether the second level of subsampling introduces recall degradation  

Does this sound right? Please let me know after the random subsampling with premutation PR is merged so I can start making changes to this PR